File: java/socketd/src/main/java/org/noear/socketd/core/impl/SessionDefault.java
package org.noear.socketd.core.impl;

import org.noear.socketd.exception.SocketdException;
import org.noear.socketd.exception.SocketdTimeoutException;
import org.noear.socketd.core.*;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.function.Consumer;

/**
 * 会话默认实现
 *
 * @author noear
 * @since 2.0
 */
public class SessionDefault extends SessionBase implements Session {
    public SessionDefault(Channel channel) {
        super(channel);
    }

    /**
     * 是否有效
     */
    @Override
    public boolean isValid() {
        return channel.isValid();
    }

    /**
     * 获取远程地址
     */
    @Override
    public InetSocketAddress getRemoteAddress() throws IOException {
        return channel.getRemoteAddress();
    }

    /**
     * 获取本地地址
     */
    @Override
    public InetSocketAddress getLocalAddress() throws IOException {
        return channel.getLocalAddress();
    }

    /**
     * 获取握手信息
     */
    @Override
    public Handshaker getHandshaker() {
        return channel.getHandshaker();
    }

    /**
     * 发送 Ping
     */
    @Override
    public void sendPing() throws IOException {
        channel.sendPing();
    }

    /**
     * 发送
     */
    @Override
    public void send(String topic, Entity content) throws IOException {
        Message message = new MessageDefault().key(generateKey()).topic(topic).entity(content);

        channel.send(new Frame(Flag.Message, message), null);
    }


    /**
     * 发送并请求（限为一次答复；指定超时）
     *
     * @param topic   主题
     * @param content 内容
     * @param timeout 超时（毫秒）
     */
    @Override
    public Entity sendAndRequest(String topic, Entity content, long timeout) throws IOException {
        //背压控制
        if (channel.getRequests().get() > channel.getConfig().getMaxRequests()) {
            throw new SocketdException("Sending too many requests: " + channel.getRequests().get());
        } else {
            channel.getRequests().incrementAndGet();
        }

        try {

            Message message = new MessageDefault().key(generateKey()).topic(topic).entity(content);

            CompletableFuture<Entity> future = new CompletableFuture<>();
            channel.send(new Frame(Flag.Request, message), new AcceptorRequest(future));
            try {
                return future.get(timeout, TimeUnit.MILLISECONDS);
            } catch (TimeoutException e) {
                throw new SocketdTimeoutException("Request timeout: " + topic);
            } catch (Throwable e) {
                throw new SocketdException(e);
            }
        } finally {
            channel.getRequests().decrementAndGet();
        }
    }

    /**
     * 发送并订阅（答复结束之前，不限答复次数）
     *
     * @param topic    主题
     * @param content  内容
     * @param consumer 回调消费者
     */
    @Override
    public void sendAndSubscribe(String topic, Entity content, Consumer<Entity> consumer) throws IOException {
        Message message = new MessageDefault().key(generateKey()).topic(topic).entity(content);
        channel.send(new Frame(Flag.Subscribe, message), new AcceptorSubscribe(consumer));
    }

    /**
     * 答复
     *
     * @param from    来源消息
     * @param content 内容
     */
    @Override
    public void reply(Message from, Entity content) throws IOException {
        channel.send(new Frame(Flag.Reply, new MessageDefault().key(from.getKey()).entity(content)), null);
    }

    /**
     * 答复并结束（即最后一次答复）
     *
     * @param from    来源消息
     * @param content 内容
     */
    @Override
    public void replyEnd(Message from, Entity content) throws IOException {
        channel.send(new Frame(Flag.ReplyEnd, new MessageDefault().key(from.getKey()).entity(content)), null);
    }

    @Override
    public void close() throws IOException {
        channel.close();
    }
}

File: python/socketd/socketd/core/SessionDefault.py
import asyncio

from abc import ABC
from typing import Callable, Awaitable, Any, Coroutine

from .SessionBase import SessionBase
from .Channel import Channel
from .Handshake import Handshake
from .module.Entity import Entity
from .module.Message import Message
from .module.Frame import Frame
from .Costants import Flag
from .module.MessageDefault import MessageDefault
from ..exception.SocketdExecption import SocketDException
from ..transport.core.CompletableFuture import CompletableFuture
from ..transport.core.StreamRequest import StreamRequest
from ..transport.core.StreamSubscribe import StreamSubscribe


class SessionDefault(SessionBase, ABC):

    def __init__(self, channel: Channel):
        super().__init__(channel)
        self.__loop = asyncio.get_event_loop()

    def is_valid(self) -> bool:
        return self.channel.is_valid()

    def get_remote_address(self) -> str:
        return self.channel.get_remote_address()

    def get_local_address(self) -> str:
        return self.channel.get_local_address()

    def get_handshake(self) -> Handshake:
        return self.channel.get_handshake()

    async def send_ping(self):
        await self.channel.send_ping()

    async def send(self, topic: str, content: Entity):
        message = MessageDefault().set_sid(self.generate_id()).set_event(topic).set_entity(content)
        await self.channel.send(Frame(Flag.Message, message), None)

    async def send_and_request(self, event: str, content: Entity,
                               timeout: int = 100) -> Entity:

        if timeout < 100:
            timeout = self.channel.get_config().get_reply_timeout()
        future: CompletableFuture[Entity] = CompletableFuture()
        message = MessageDefault().set_sid(self.generate_id()).set_event(event).set_entity(content)
        try:
            await self.channel.send(Frame(Flag.Request, message),
                                    StreamRequest(message.get_sid(), timeout, future))
            return await future.get(timeout)
        except asyncio.TimeoutError as e:
            if self.channel.is_valid():
                raise SocketDException(f"Request reply timeout>{timeout} "
                                f"sessionId={self.channel.get_session().get_session_id()} "
                                f"event={event} sid={message.get_sid()}")
            else:
                raise SocketDException(f"This channel is closed sessionId={self.channel.get_session().get_session_id()} "
                                f"event={event} sid={message.get_sid()}")
        except Exception as e:
            raise e
        finally:
            self.channel.remove_acceptor(message.get_sid())

    async def send_stream_and_request(self, event: str, content: Entity,
                                      consumer: Callable[[Entity], Awaitable[Any]] | Coroutine[Entity, Any, None],
                                      timeout: int):
        message = MessageDefault().set_sid(self.generate_id()).set_event(event).set_entity(content)
        future: CompletableFuture[Entity] = CompletableFuture()
        try:
            if asyncio.iscoroutinefunction(consumer):
                await consumer(content)
            else:
                self.channel.get_config().get_executor().submit(fn=consumer, args=(content,))
        except Exception as e:
            self.channel.on_error(e)
        streamAcceptor = StreamRequest(message.get_sid(), timeout, future)
        await self.channel.send(Frame(Flag.Request, message), streamAcceptor)
        return streamAcceptor

    async def send_and_subscribe(self, event: str, content: Entity, consumer: Callable[[Entity], Any], timeout: int):
        message = MessageDefault().set_sid(self.generate_id()).set_event(event).set_entity(content)
        streamAcceptor = StreamSubscribe(message.get_sid(), timeout, consumer)
        await self.channel.send(Frame(Flag.Subscribe, message), streamAcceptor)
        return streamAcceptor

    async def reply(self, from_msg: Message, content: Entity):
        await self.channel.send(Frame(Flag.Reply,
                                      MessageDefault()
                                      .set_sid(from_msg.get_sid())
                                      .set_event(from_msg.get_event())
                                      .set_entity(content)), None)

    async def reply_end(self, from_msg: Message, content: Entity):
        await self.channel.send(Frame(Flag.ReplyEnd,
                                      MessageDefault()
                                      .set_sid(from_msg.get_sid())
                                      .set_event(from_msg.get_event())
                                      .set_entity(content)), None)

    async def close(self):
        if self.channel.is_valid():
            await self.channel.send_close()
            await self.channel.close()
